pennylane/gradients/hadamard_gradient.py

```diff
@@ -309,218 +290,122 @@ def hadamard_grad(
-                elif p_idx == 1:
-                    # Apply additional rotations that effectively move the generator to the middle of Rot
-                    ops_to_add_before = [
-                        qml.RZ(-trainable_op.data[2], wires=trainable_op.wires),
-                        qml.RX(np.pi / 2, wires=trainable_op.wires),
-                    ]
-                    ops_to_trainable_op.extend(ops_to_add_before)
-
-                    ops_to_add_after = [
-                        qml.RX(-np.pi / 2, wires=trainable_op.wires),
-                        qml.RZ(trainable_op.data[2], wires=trainable_op.wires),
-                    ]
-                    ops_after_trainable_op = ops_to_add_after + ops_after_trainable_op
-
             ctrl_gen = [qml.ctrl(gen, control=aux_wire)]
             hadamard = [qml.Hadamard(wires=aux_wire)]
             ops = ops_to_trainable_op + hadamard + ctrl_gen + hadamard + ops_after_trainable_op
 
-            measurements = []
-            # Add the Y measurement on the aux qubit
-            for m in tape.measurements:
-                if m.obs:
-                    obs_new = [m.obs]
-                else:
-                    m_wires = m.wires if len(m.wires) > 0 else tape.wires
-                    obs_new = [qml.Z(i) for i in m_wires]
-
-                obs_new.append(qml.Y(aux_wire))
-                obs_new = qml.prod(*obs_new)
-
-                if isinstance(m, qml.measurements.ExpectationMP):
-                    measurements.append(qml.expval(op=obs_new))
-                else:
-                    measurements.append(qml.probs(op=obs_new))
-
-            new_tape = qml.tape.QuantumScript(ops=ops, measurements=measurements, shots=tape.shots)
-            _rotations, _measurements = qml.tape.tape.rotations_and_diagonal_measurements(new_tape)
-            new_ops = new_tape.operations + _rotations
-            new_tape = qml.tape.QuantumScript(
-                new_ops,
-                _measurements,
-                shots=new_tape.shots,
-                trainable_params=new_tape.trainable_params,
-            )
-
-            num_tape += 1
+            measurements = (_new_measurement(mp, aux_wire, tape.wires) for mp in tape.measurements)
+            new_tape = qml.tape.QuantumScript(ops, measurements, shots=tape.shots)
 
             g_tapes.append(new_tape)
 
-        gradient_data.append(num_tape)
+    return g_tapes, partial(
+        processing_fn, coeffs=coeffs, tape=tape, generators_per_parameter=generators_per_parameter
+    )
+
+
+def _new_measurement(mp, aux_wire, all_wires: qml.wires.Wires):
+    obs = mp.obs or qml.prod(*(qml.Z(w) for w in mp.wires or all_wires))
+    new_obs = qml.simplify(obs @ qml.Y(aux_wire))
+    return type(mp)(obs=new_obs)
 
-    multi_measurements = len(tape.measurements) > 1
-    multi_params = len(tape.trainable_params) > 1
+
+def _get_pauli_generators(trainable_op):
```

Comment:
Now rely on the `Operator` interface to have a generic algorithm for getting the generator as a linear combination of pauli's, instead of having to have a bespoke registration for each operator. This allows to differentiate any time evolution op.